rustls/src/server/tls12.rs

```diff
@@ -46,40 +47,50 @@ mod client_hello {
         Random, ServerExtensionsInput, ServerHelloPayload, ServerKeyExchange,
         ServerKeyExchangeParams, ServerKeyExchangePayload,
     };
+    use crate::sealed::Sealed;
     use crate::sign;
     use crate::verify::DigitallySignedStruct;
 
-    pub(in crate::server) struct CompleteClientHelloHandling {
-        pub(in crate::server) config: Arc<ServerConfig>,
-        pub(in crate::server) transcript: HandshakeHash,
-        pub(in crate::server) session_id: SessionId,
-        pub(in crate::server) suite: &'static Tls12CipherSuite,
-        pub(in crate::server) using_ems: bool,
-        pub(in crate::server) randoms: ConnectionRandoms,
-        pub(in crate::server) send_ticket: bool,
-        pub(in crate::server) extra_exts: ServerExtensionsInput<'static>,
+    pub(crate) static TLS12_HANDLER: &dyn Tls12Handler = &Handler;
+
+    pub(crate) trait Tls12Handler: Sealed + Send + Sync + core::fmt::Debug {
+        fn handle_client_hello(
+            &self,
+            cx: &mut ServerContext<'_>,
+            server_key: ActiveCertifiedKey<'_>,
+            chm: &Message<'_>,
+            client_hello: &ClientHelloPayload,
+            selected_kxg: &'static dyn SupportedKxGroup,
+            sigschemes_ext: Vec<SignatureScheme>,
+            tls13_enabled: bool,
+            cch: CompleteClientHelloHandling,
+        ) -> hs::NextStateOrError<'static>;
     }
 
-    impl CompleteClientHelloHandling {
-        pub(in crate::server) fn handle_client_hello(
-            mut self,
+    #[derive(Debug)]
+    struct Handler;
+
+    impl Tls12Handler for Handler {
+        fn handle_client_hello(
+            &self,
             cx: &mut ServerContext<'_>,
             server_key: ActiveCertifiedKey<'_>,
             chm: &Message<'_>,
             client_hello: &ClientHelloPayload,
             selected_kxg: &'static dyn SupportedKxGroup,
             sigschemes_ext: Vec<SignatureScheme>,
             tls13_enabled: bool,
+            mut cch: CompleteClientHelloHandling,
```

Comment:
Nits: same here about isolating the `self` -> `cch` renames. Also suggest moving `cch` up front in the argument list.